src/main/java/edu/wpi/grip/core/serialization/SocketConverter.java

```diff
@@ -0,0 +1,170 @@
+            }
+
+            writer.endNode();
+        } catch (ConversionException e) {
+            throw new ConversionException("Error serializing socket: " + socket, e);
+        }
+    }
+
+    @Override
+    public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext context) {
+        try {
+            reader.moveDown();
+
+            final String nodeName = reader.getNodeName();
+
+            Socket.Direction direction;
+            if (nodeName.equals(mapper.serializedClass(InputSocket.class))) {
+                direction = Socket.Direction.INPUT;
+            } else if (nodeName.equals(mapper.serializedClass(OutputSocket.class))) {
+                direction = Socket.Direction.OUTPUT;
+            } else {
+                throw new IllegalArgumentException("Unexpected socket node name: " + nodeName);
+            }
+
+            // Look up the socket using the saved indexes.  Serializing sockets this way makes it so different things
+            // (such as connections) can reference sockets in the pipeline.
+            Socket socket;
+            if (reader.getAttribute(STEP_ATTRIBUTE) != null) {
+                final int stepIndex = Integer.valueOf(reader.getAttribute(STEP_ATTRIBUTE));
+                final int socketIndex = Integer.valueOf(reader.getAttribute(SOCKET_ATTRIBUTE));
+
+                final Step step = pipeline.getSteps().get(stepIndex);
+                socket = direction == Socket.Direction.INPUT ?
+                        step.getInputSockets()[socketIndex] : step.getOutputSockets()[socketIndex];
+            } else if (reader.getAttribute(SOURCE_ATTRIBUTE) != null) {
+                final int sourceIndex = Integer.valueOf(reader.getAttribute(SOURCE_ATTRIBUTE));
+                final int socketIndex = Integer.valueOf(reader.getAttribute(SOCKET_ATTRIBUTE));
+
+                final Source source = pipeline.getSources().get(sourceIndex);
+                socket = source.getOutputSockets()[socketIndex];
+            } else {
+                throw new ConversionException("Sockets must have either a step or source attribute");
+            }
+
+            if (socket.getDirection() == Socket.Direction.OUTPUT) {
+                ((OutputSocket) socket).setPublished(Boolean.valueOf(reader.getAttribute(PUBLISHED_ATTRIBUTE)));
+                ((OutputSocket) socket).setPreviewed(Boolean.valueOf(reader.getAttribute(PREVIEWED_ATTRIBUTE)));
+            }
+
+            if (reader.hasMoreChildren()) {
+                reader.moveDown();
+                if (reader.getNodeName().equals("value")) {
+                    socket.setValue(context.convertAnother(socket, getDeserializedType(socket)));
+                } else {
+                    throw new IllegalArgumentException("Unexpected node in socket: " + reader.getNodeName());
+                }
+                reader.moveUp();
+            }
+
+            reader.moveUp();
+
+            return socket;
+        } catch (Exception e) {
```

Comment:
This is intentionally so that any exception is rethrown as a `ConversionException`.